tcbuilder/backend/bundle.py

```diff
@@ -487,6 +485,13 @@ def login_to_registries(client, logins):
         client.login(username, password, registry=registry)
 
 
+def check_double_dollar_signs(string):
+    if "$$" in string:
```

Comment:
Checking the documentation on [Interpolation](https://docs.docker.com/reference/compose-file/interpolation/) I notice that the feature works in a way that differs from what this stretch of code assumes (that the substitutions are performed before parsing the file as yaml (i.e. in the raw text)). In fact, the docs says "Interpolation applies only to YAML values, not to keys".

Another complication I found is what is said [here](https://docs.docker.com/compose/environment-variables/variable-interpolation/#interpolation-syntax): "Interpolation is applied for unquoted and double-quoted values.", i.e. apparently not to single quoted strings.

In light of the above, the question now becomes: do we want to handle the extra complexity to ensure backwards compatibility? If the answer is yes, then we'd need to parse the yaml and then perform the replacement `$$` => `$` only on unquoted and double quoted string **values**. Opinions?

Reply:
Parsing the YAML contents to replace `$$` with `$` in string _values_  should be doable without much effort.

However, the main challenge that I see with this is to only select unquoted and double quoted strings, especially after parsing the YAML file, as all of those would be converted to Python strings, with no easy way to tell whether they were originally unquoted, single or double quoted in the YAML file.

While this unquoted/double quoted requirement for interpolation is specified in the Compose V2 documentation, I couldn't find it on the legacy Compose V1 documentation: https://github.com/docker/compose/blob/v1/docs/Compose%20file%20reference%20(legacy)/version-3.md#variable-substitution

Reply:
That's a good point.

Well, in the end what matters for us is if the methods we used before did the substitution or not. Performing a quick test it seems it does the substitution on all value strings (but not the keys):

E.g. entering this to `tcb bundle`:

```
services:
  hello-world:
    image: hello-world:latest
    environment:
      $var: $HOME
      HOME1: $HOME
      HOME2: "$HOME"
      HOME3: '$HOME'
      HOME4:
        >-
          $HOME
      HOME5: |
          $HOME
      HOME6: |
          \$HOME
```

gives this output:

```
services:
  hello-world:
    environment:
      $var: /root
      HOME1: /root
      HOME2: /root
      HOME3: /root
      HOME4: /root
      HOME5: '/root

        '
      HOME6: '\/root

        '
    image: hello-world@sha256:91fb4b041da273d5a3273b6d587d62d518300a6ad268b28628f74997b93171b2
version: '3.9'
```

So, it should be enough to make the substitutions on all string values...

Reply:
That's great to know, thanks for the tests!

Aside from this, the Compose V2 docs also mention the fact that a user may want to keep the double-dollar sign if they don't want environment variables being processed by Compose:
```
web:
  build: .
  command: "$$VAR_NOT_INTERPOLATED_BY_COMPOSE"
```

To cover this use case I can also add a switch to the bundle command (and its respective option on tcbuild.yaml) to disable the replacement, something like `--keep-double-dollar-sign`.

Reply:
Once we get rid of the workaround documentation from our docs, then any users using the `$$` in their compose would likely be following the Compose documentation and them their expectation would be that TCB doesn't itself touch the file.

So, yes I think it makes sense.